### PR TITLE
Build runc.riscv64 

### PR DESCRIPTION
**Pull request by AkihiroSuda**
The first commit bumps up libseccomp-golang (https://github.com/seccomp/libseccomp-golang/compare/3879420cc921...e214ef109e10)  to support `SCMP_ARCH_RISCV64` (https://github.com/seccomp/libseccomp-golang/commit/e947d821936c4eb924b41542e4cd6b1144812f79).

The second commit updates the build scripts to build `runc.riscv64`.

- - -

Tested on Ubuntu 22.04 (kernel 5.15) on `qemu-system-riscv64 -M virt -cpu rv64` (QEMU 6.2, https://github.com/lima-vm/lima/pull/827)

```console
# uname -a
Linux lima-riscv64 5.15.0-1007-generic #7-Ubuntu SMP Tue Apr 12 16:07:30 UTC 2022 riscv64 riscv64 riscv64 GNU/Linux
# runc --version
runc version 1.1.0+dev
commit: v1.1.0-161-gc0219678
spec: 1.0.2-dev
go: go1.18.1
libseccomp: 2.5.3
# mkdir rootfs
# curl -fsSL https://dl-cdn.alpinelinux.org/alpine/edge/releases/riscv64/alpine-minirootfs-20220316-riscv64.tar.gz | tar Cxz ./rootfs
# cat <<EOF >config.json
{
  "ociVersion": "1.0.2-dev",
  "process": {
    "terminal": true,
    "user": { "uid": 0, "gid": 0 },
    "args": [ "sh" ],
    "env": [ "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", "TERM=xterm" ],
    "cwd": "/",
    "capabilities": {
      "bounding": [ "CAP_AUDIT_WRITE", "CAP_KILL", "CAP_NET_BIND_SERVICE" ],
      "effective": [ "CAP_AUDIT_WRITE", "CAP_KILL", "CAP_NET_BIND_SERVICE" ],
      "inheritable": [ "CAP_AUDIT_WRITE", "CAP_KILL", "CAP_NET_BIND_SERVICE" ],
      "permitted": [ "CAP_AUDIT_WRITE", "CAP_KILL", "CAP_NET_BIND_SERVICE" ],
      "ambient": [ "CAP_AUDIT_WRITE", "CAP_KILL", "CAP_NET_BIND_SERVICE" ]
    },
    "rlimits": [ { "type": "RLIMIT_NOFILE", "hard": 1024, "soft": 1024 } ],
    "noNewPrivileges": true
  },
  "root": { "path": "rootfs", "readonly":false },
  "hostname": "runc",
  "mounts": [
    { "destination": "/proc", "type": "proc", "source": "proc" },
    { "destination": "/dev", "type": "tmpfs", "source": "tmpfs", "options": [ "nosuid", "strictatime", "mode=755", "size=65536k" ] },
    { "destination": "/dev/pts", "type": "devpts", "source": "devpts", "options": [ "nosuid", "noexec", "newinstance", "ptmxmode=0666", "mode=0620", "gid=5" ] },
    { "destination": "/dev/shm", "type": "tmpfs", "source": "shm", "options": [ "nosuid", "noexec", "nodev", "mode=1777", "size=65536k" ] },
    { "destination": "/dev/mqueue", "type": "mqueue", "source": "mqueue", "options": [ "nosuid", "noexec", "nodev" ] },
    { "destination": "/sys", "type": "sysfs", "source": "sysfs", "options": [ "nosuid", "noexec", "nodev", "ro" ] },
    { "destination": "/sys/fs/cgroup", "type": "cgroup", "source": "cgroup", "options": [ "nosuid", "noexec", "nodev", "relatime", "ro" ] }
  ],
  "linux": {
    "resources": { "devices": [ { "allow": false, "access": "rwm" } ] },
    "namespaces": [ { "type": "pid" }, { "type": "network" }, { "type": "ipc" }, { "type": "uts" }, { "type": "mount" }, { "type": "cgroup" } ],
    "maskedPaths": [ "/proc/acpi", "/proc/asound", "/proc/kcore", "/proc/keys", "/proc/latency_stats", "/proc/timer_list", "/proc/timer_stats", "/proc/sched_debug", "/sys/firmware", "/proc/scsi" ],
    "readonlyPaths": [ "/proc/bus", "/proc/fs", "/proc/irq", "/proc/sys", "/proc/sysrq-trigger" ],
    "seccomp": {
      "defaultAction": "SCMP_ACT_ALLOW",
      "architectures": [ "SCMP_ARCH_RISCV64" ],
      "syscalls": [ { "names": [ "unshare" ], "action": "SCMP_ACT_ERRNO", "errnoRet": 42 } ]
    }
  }
}
EOF
# runc run foo
/ # uname -a
Linux runc 5.15.0-1007-generic #7-Ubuntu SMP Tue Apr 12 16:07:30 UTC 2022 riscv64 Linux
/ # unshare
unshare: unshare(0x0): No message of desired type
/ # : unshare is failing with ENOMSG (errno 42) as expected
/ # 
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -22,6 +23,8 @@ RUN KEYFILE=/usr/share/keyrings/criu-repo-keyring.gpg; \
         crossbuild-essential-armhf \
         crossbuild-essential-ppc64el \
         crossbuild-essential-s390x \
+        binutils-riscv64-linux-gnu cpp-riscv64-linux-gnu gcc-riscv64-linux-gnu \
+        libc6-riscv64-cross libc6-dev-riscv64-cross linux-libc-dev-riscv64-cross \
```
Comment (thaJeztah) [inline Dockerfile:27]: Perhaps put the comment before this section, and make it a "group" of things to install to make clear "these are the bits related to risc-v, by adding an "empty" line before/after e.g.;

```dockerfile
RUN KEYFILE=/usr/share/keyrings/criu-repo-keyring.gpg; \
    wget -nv $CRIU_REPO/Release.key -O- | gpg --dearmor > "$KEYFILE" \
    && echo "deb [signed-by=$KEYFILE] $CRIU_REPO/ /" > /etc/apt/sources.list.d/criu.list \
    && dpkg --add-architecture armel \
    && dpkg --add-architecture armhf \
    && dpkg --add-architecture arm64 \
    && dpkg --add-architecture ppc64el \
    && apt-get update \
    && apt-get install -y --no-install-recommends \
        build-essential \
        criu \
        crossbuild-essential-arm64 \
        crossbuild-essential-armel \
        crossbuild-essential-armhf \
        crossbuild-essential-ppc64el \
        crossbuild-essential-s390x \
        \
# no "crossbuild-essential-riscv64" package yet: https://wiki.debian.org/RISC-V
        binutils-riscv64-linux-gnu \
        cpp-riscv64-linux-gnu \
        gcc-riscv64-linux-gnu \
        libc6-dev-riscv64-cross \
        libc6-riscv64-cross \
        linux-libc-dev-riscv64-cross \
        \
        curl \
        gawk \
        gcc \
        gperf \
        iptables \
        jq \
        kmod \
        pkg-config \
        python3-minimal \
        sshfs \
        sudo \
        uidmap \
    && apt-get clean \
    && rm -rf /var/cache/apt /var/lib/apt/lists/* /etc/apt/sources.list.d/*.list
```

File: Dockerfile [hunk 2]
```diff
@@ -56,7 +59,7 @@ RUN cd /tmp \
 ARG LIBSECCOMP_VERSION
 COPY script/* /tmp/script/
 RUN mkdir -p /opt/libseccomp \
-    && /tmp/script/seccomp.sh "$LIBSECCOMP_VERSION" /opt/libseccomp arm64 armel armhf ppc64le s390x
+    && /tmp/script/seccomp.sh "$LIBSECCOMP_VERSION" /opt/libseccomp arm64 armel armhf ppc64le s390x riscv64
```
Comment (thaJeztah) [inline Dockerfile:62]: Looks like the list was intended to be sorted alphabetically;

```suggestion
    && /tmp/script/seccomp.sh "$LIBSECCOMP_VERSION" /opt/libseccomp arm64 armel armhf ppc64le riscv64 s390x
```

File: Makefile [hunk 3]
```diff
@@ -40,7 +40,7 @@ recvtty sd-helper seccompagent:
 static:
 	$(GO_BUILD_STATIC) -o runc .
 
-releaseall: RELEASE_ARGS := "-a arm64 -a armel -a armhf -a ppc64le -a s390x"
+releaseall: RELEASE_ARGS := "-a arm64 -a armel -a armhf -a ppc64le -a s390x -a riscv64"
```
Comment (thaJeztah) [inline Makefile:43]: same here

```suggestion
releaseall: RELEASE_ARGS := "-a arm64 -a armel -a armhf -a ppc64le -a riscv64 -a s390x"
```

File: libcontainer/seccomp/config.go [hunk 4]
```diff
@@ -66,6 +66,7 @@ var archs = map[string]string{
 	"SCMP_ARCH_PPC64LE":     "ppc64le",
 	"SCMP_ARCH_S390":        "s390",
 	"SCMP_ARCH_S390X":       "s390x",
+	"SCMP_ARCH_RISCV64":     "riscv64",
```
Comment (thaJeztah) [inline libcontainer/seccomp/config.go:69]: We should probably sort this list as well, but looks like it's already a bit "unsorted"; perhaps move it before the `s390x` variants though (somewhat more consistent with the other lists)

File: script/lib.sh [hunk 5]
```diff
@@ -26,6 +26,9 @@ function set_cross_vars() {
 	s390x)
 		HOST=s390x-linux-gnu
 		;;
+	riscv64)
+		HOST=riscv64-linux-gnu
+		;;
```
Comment (thaJeztah) [inline script/lib.sh:31]:
```suggestion
	riscv64)
		HOST=riscv64-linux-gnu
		;;
	s390x)
		HOST=s390x-linux-gnu
		;;
```

**Review by thaJeztah — COMMENTED** [review]
thx! left some nits for consistency 😅

File: Dockerfile (hunk 1 above)
Comment (crazy-max) [reply Dockerfile:27]: Maybe [xx](https://github.com/tonistiigi/xx/) could be useful so you don't need to install all cross packages for each platform but relies on platform args in global scope. Ofc it depends if you're using buildx/BuildKit to build. See https://github.com/moby/moby/pull/43529 and https://github.com/crazy-max/moby/blob/ed158db3ad1c31300a59ccd0e70c52384d5d7308/Dockerfile.cross#L108-L123

File: script/lib.sh (hunk 5 above)
Comment (AkihiroSuda) [reply script/lib.sh]: done

File: libcontainer/seccomp/config.go (hunk 4 above)
Comment (AkihiroSuda) [reply libcontainer/seccomp/config.go]: done

File: Makefile (hunk 3 above)
Comment (AkihiroSuda) [reply Makefile]: done

File: Dockerfile (hunk 2 above)
Comment (AkihiroSuda) [reply Dockerfile]: done

File: Dockerfile (hunk 1 above)
Comment (AkihiroSuda) [reply Dockerfile]: done.

Not sure we are allowed to use buildx/buildkit here.

File: libcontainer/seccomp/config.go [hunk 6]
```diff
@@ -64,6 +64,7 @@ var archs = map[string]string{
 	"SCMP_ARCH_PPC":         "ppc",
 	"SCMP_ARCH_PPC64":       "ppc64",
 	"SCMP_ARCH_PPC64LE":     "ppc64le",
+	"SCMP_ARCH_RISCV64":     "riscv64",
```
Comment (thaJeztah) [inline libcontainer/seccomp/config.go:67]: oh! think this one ended up in the wrong commit, or was it intentional to be part of the vendor update?

**Review by thaJeztah — APPROVED** [review]
LGTM (left one question)

File: libcontainer/seccomp/config.go (hunk 6 above)
Comment (AkihiroSuda) [reply libcontainer/seccomp/config.go:67]: Yes. Tied with the new version of libseccomp-golang.

**Comment by crazy-max** [comment]
@AkihiroSuda I can test on my side too if you want, I have an HiFive Unmatched. Let me know.

**Comment by AkihiroSuda** [comment]
@crazy-max Thanks, testing on a real machine is highly appreciated 🙏 .
I only tested it on `qemu-system-riscv64 -M virt -cpu rv64`.

**Comment by crazy-max** [comment]
@AkihiroSuda 

```
$ uname -a
Linux sifive 5.11.0-1022-generic #23-Ubuntu SMP Wed Oct 20 17:18:01 UTC 2021 riscv64 riscv64 riscv64 GNU/Linux
$ git clone https://github.com/AkihiroSuda/runc
$ git checkout riscv64
$ cd runc/
$ make
$ file runc
runc: ELF 64-bit LSB pie executable, UCB RISC-V, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-riscv64-lp64d.so.1, BuildID[sha1]=024bc8788be092ecea2a41b1df7e6f9841e71fd8, for GNU/Linux 4.15.0, not stripped
$ runc --version
runc version 1.0.1-0ubuntu2~21.04.1
spec: 1.0.2-dev
go: go1.16.2
libseccomp: 2.5.1
```

Let me know if this is enough to test its behavior.

**Comment by kolyshkin** [comment]
Superceded by https://github.com/opencontainers/runc/pull/3446

**Comment by crazy-max** [comment]
@AkihiroSuda 

```
$ runc run foo
/ # uname -a
Linux runc 5.11.0-1022-generic #23-Ubuntu SMP Wed Oct 20 17:18:01 UTC 2021 riscv64 Linux
/ # unshare
unshare: unshare(0x0): No message of desired type
/ #
```

**Comment by AkihiroSuda** [comment]
Thank you @crazy-max , it is working as expected

**Pull request closed without merging**